### PR TITLE
Smiles data reader

### PR DESCRIPTION
**Pull request by davidHysom**
Data reader for SMILES data

### CODE REVIEW

File: applications/ATOM/data_config.json [hunk 1]
```diff
@@ -1,7 +1,5 @@
 {
-
-  "pad_index": 28,
-  "max_seq_len": 57,
-  "data_path": "/p/lustre2/brainusr/datasets/zinc/moses_zinc_train250K.npy"
-
+  "pad_index": 34,
```
Comment (samadejacobs) [inline applications/ATOM/data_config.json:2]: Please reverse the changes here, lets keep the default value.

File: applications/ATOM/smiles_data_reader.prototext [hunk 2]
```diff
@@ -0,0 +1,14 @@
+data_reader {
+  reader {
+    name: "smiles"
+    role: "train"
+    shuffle: true
+    percent_of_data_to_use: 1.0
+    data_filedir: "/p/lustre2/brainusr/datasets/enamine"
+    data_filename: "all2018q1-2REAL680SMILES.smiles"
+
+    #data_filename: "good.csv"
```
Comment (samadejacobs) [inline applications/ATOM/smiles_data_reader.prototext:10]: Do we need lines 10-12?

File: applications/ATOM/tools/compute_vocab.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
+#include <iostream>
+#include <string>
```
Comment (samadejacobs) [inline applications/ATOM/tools/compute_vocab.cpp:2]: I suggest we rename _tools_ directory as _utils_ directory. I have third party ("sophisticated") python scripts as well. Unlike this scripts that is standalone, mine are called from python frontend so I place them in utils directory. It is fine with me if choose to retain _tools_ directory, just a comment.

File: include/lbann/layers/learning/fully_connected.hpp [hunk 4]
```diff
@@ -212,7 +212,7 @@ class fully_connected_layer : public learning_layer<TensorDataType> {
 
     // Initialize freeze state
     for (auto&& w : this->get_data_type_weights()) {
-      if (this->m_frozen) {
+      if (this->is_frozen()) {
         w->freeze();
```
Comment (samadejacobs) [inline include/lbann/layers/learning/fully_connected.hpp:216]: Note, this file may have nothing to do with this PR.

File: applications/ATOM/data_config.json (hunk 1 above)
Comment (davidHysom) [reply applications/ATOM/data_config.json]: fixed.

File: applications/ATOM/smiles_data_reader.prototext (hunk 2 above)
Comment (davidHysom) [reply applications/ATOM/smiles_data_reader.prototext]: fixed.

Comment (davidHysom) [reply applications/ATOM/smiles_data_reader.prototext]: deleted those lines

File: include/lbann/layers/learning/fully_connected.hpp (hunk 4 above)
Comment (davidHysom) [reply include/lbann/layers/learning/fully_connected.hpp]: changed back to this->m_frozen, which agrees with the head of develop. I don't know where the change came from -- nothing I did intentionally.

File: src/layers/layer.cpp [hunk 5]
```diff
@@ -358,8 +358,7 @@ void Layer::unfreeze() {
 bool Layer::is_frozen() const {
   for(auto& w : get_weights()) {
     if (w->is_frozen() != m_frozen) {
-      LBANN_ERROR("layer ", get_name(), " and weight ", w->get_name(), \
-                  " of it are inconsistently frozen");
+      LBANN_ERROR("layer and weights of them are inconsistently frozen");
```
[the remark this answers is not in the document]
Comment (davidHysom) [reply src/layers/layer.cpp:361]: This is not part of this PR, and nothing that I did. Sam saw there was an error here -- I've no idea where it came from, or how this piece of code is supposed to read.

**Comment by davidHysom** [comment]
Pertinent changes/additions/modifications were copied to PR #1533; that PR started as an up-to-date clone of my fork, hence, avoids changes to layer.hpp, etc, that somehow snuck in. Closing this PR, since #1533 has been merged.

**Pull request closed without merging**